### ISSUE TITLE
better support for postgres array/hstore types

### ISSUE THREAD

**Issue by 0x777** [hasura/graphql-engine#348]
Currently the input for an array is a literal, i.e for a column of type `text[]` the input value has to be a string as follows: `"{hello,world}"`. We should do the appropriate mapping to json arrays.

**Comment by 0x777** [comment]
via Pum on discord

**Comment by splodingsocks** [comment]
This works when the literal is typed out in the body of the query, but I can't even get the string syntax to work when the value is passed in as a variable.

**Comment by splodingsocks** [comment]
Actually, turns out that variables work as long as their type is `_text` and not `String`.

**Comment by 0x777** [comment]
We should also be supporting the common operators on arrays like we did for json/jsonb.

**Comment by jafri** [comment]
Most importantly, would like to see `_contains` support on arrays

**Comment by coco98** [comment]
Till we add this, as a workaround, you can use JSONB and store JSON encoded
arrays. We have a support for a bunch of operators including  some that
have contains type functionality.

**Comment by jafri** [comment]
@coco98  I have seen the `@>` contains support for JSONB, the problem is that primitive arrays provided statistics, and are thus more efficient in queries.

**Comment by cj** [comment]
I use this a lot with Rails https://edgeguides.rubyonrails.org/active_record_postgresql.html#array

**Comment by yassinebridi** [comment]
Is there an update on this.
it's getting really hard updating the arrays, no mapping...
we always need to do some hacks to accomplish those simple operations.

**Comment by dsandip** [comment]
Hi @yassinebridi, https://github.com/hasura/graphql-engine/pull/2243 had to be aborted to handle some edge cases. We'll be picking this up again in a couple of weeks.

**Comment by xxRockOnxx** [comment]
What can we do for now instead... without resorting to implementing custom server just to have a single endpoint for this exact query?

If there's none, then I guess that's how it should be.

**Comment by gunar-gessner-qatalog** [comment]
fyi just knowing that the right format for `text[]` is `{hello,world}` solved my problem (obvious in retrospect) — consider adding it to docs

**Comment by gunar** [comment]
fyi just knowing that the right format for `text[]` is `{hello,world}` solved my problem (obvious in retrospect) — consider adding it to docs

**Comment by marionschleifer** [comment]
@gunar you can find this in our docs [here](https://docs.hasura.io/1.0/graphql/manual/mutations/insert.html#insert-an-object-with-an-array-field) 🙂

**Comment by gunar** [comment]
@marionschleifer that's great, thank you! — your docs are impeccable.

**Comment by flinz** [comment]
Still excitedly waiting for #2243 !

**Comment by JorgeCeja** [comment]
@0x777 Can `text[]` be accessed through the console?

**Comment by tirumaraiselvan** [comment]
@JorgeCeja This issue is for server. Suppose for array types in console is here: https://github.com/hasura/graphql-engine/issues/2544

**Comment by aydts** [comment]
Array Support PLZ and Thank you! I am enjoying Hasura

**Comment by AkshayCloudAnalogy** [comment]
Any update on how we can use array and what will be the type of field for it in hasura?
https://hasura.io/docs/1.0/graphql/manual/mutations/insert.html#insert-an-object-with-an-array-field
This is not working and sending back as string and also there is no text[] or _text in field types in hasura.

**Comment by cpursley** [comment]
Is it possible or planned to have the `_contains` filter type for arrays? Arrays seem to work in general but not for filtering.

**Comment by ghashi** [comment]
@JorgeCeja , @AkshayCloudAnalogy  .. there's a way to "access" the `text[]` through the console (I know, it's a hack 😅)

Here's a gif showing how:
- you can type the "[]" symbol after selecting the "text" type

![array](https://user-images.githubusercontent.com/884725/97246642-94b3fe00-17dc-11eb-8242-29a057e47d71.gif)

**Comment by rtrembecky** [comment]
I migrated a column from `jsonb` to `text[]`, defined `__text: string[]` in `codegen.yml` and generated typescript types. The type of queried data was cool, but I found out I can't insert arrays to the field because the data should be in this array literal format... Looking forward to this issue being addressed.

**Comment by bbigras** [comment]
Any progress on this? I'm asking because someone is assigned.

**Comment by danstepanov** [comment]
@rtrembecky Not sure I understand "data should be in this array literal." Is the reason the array can't be updated due to something with Hasura or the client library you use to talk to Hasura (I use React Apollo, for example)? Were you able to format the data in a way that you could update the arrray?

**Comment by rtrembecky** [comment]
@danstepanov this was nearly a year ago, so I don't remember well, but I believe this is a problem of Hasura - when you use `text[]` as a field format, you receive the data as an array, but you need to insert/update data in "array literal format" instead - `"{hello,world}"` as mentioned by OP here. this would require somehow migrating all arrays in mutations to this format.
my goal was to make the typescript types better and be more confident in our code (not to introduce clutter), so I decided to stick to `jsonb` and to hope this gets solved (and we see it never did).

**Comment by danstepanov** [comment]
Got it, thanks @rtrembecky.

The way I have gotten around this is passing in the variable as:
```
variables: { stringArray: `{ ${newArrayItem} }` }
```
within `useMutation`. And my actual query type definition is:
```
gql`
    mutation MyMutation(
        $stringArray: _text
    ) {
        REST_OF_MUTATION_HERE...
    }
```
This is definitely bonkers and I haven't yet set it up for if you want to keep the old values of `stringArray` and simply add one more item to it. But it works for now.

**Comment by factoidforrest** [comment]
This is kind of bonkers, indeed. I wonder why arrays are second class citizens in hasura and what the decision making was behind that. The fact that JSONb operators work for this use case seems like more of an accident than an intended workaround.

**Comment by John0x** [comment]
Having proper support for arrays would be great. Kinda sucks that it still isn't supported even though this issue is from 2018

**Comment by zzart** [comment]
Please make the arrays work!

**Comment by maxcan** [comment]
coming on 4 years...

**Comment by adhendo** [comment]
Bump,

Another workaround is creating a new table for your array with a foreign key to relate it where you want.

**Comment by karibertils** [comment]
Would like to see proper array support.

**Comment by denis-ryzhkov** [comment]
@rccc is right.

I loved Hasura, even created a [Python client ahasura](https://pypi.org/project/ahasura/), but then I found [TypeGraphQL-Prisma integration](https://prisma.typegraphql.com/) which looks very much like Hasura, but it is code-first and [supports scalar lists in a nice way](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-fields/working-with-scalar-lists-arrays), so now I'm trying to use it instead of Hasura.

Prisma has questionable choices too, like [in-memory `distinct` instead of DB-level `SELECT DISTINCT`](https://www.prisma.io/docs/concepts/components/prisma-client/aggregation-grouping-summarizing#distinct-under-the-hood), so I hope Hasura could prioritize its issues better to keep competition.

**Comment by SGChristianMP** [comment]
![Still waiting](https://media.giphy.com/media/FoH28ucxZFJZu/giphy.gif)

**Comment by cpursley** [comment]
Fwiw, you can use Postgres JSONB with a json array. Works pretty well for me.

**Comment by denis-ryzhkov** [comment]
@cpursley right, when you have no requirement of strict typing and validation of data inside of JSONB, it is a great data type. Until you get such requirement :)

**Comment by maxcan** [comment]
Can you use postgres check constraints to verify the type of JSONB object?

**Comment by denis-ryzhkov** [comment]
@maxcan yes, you can `CHECK (jsonb_typeof(column) = 'array')`, but to check the type of each item in this untyped array you will probably need a stored procedure, which are complex to maintain.

**Comment by zombie6888** [comment]
As workaround, you can use custom serializers, if your codegen package support it. For anyone who looking for dart/flutter solution i can share my overrides for some postgres schema specific types.

**Comment by manurana** [comment]
Supabase has [opensourced](https://supabase.com/blog/pg-jsonschema-a-postgres-extension-for-json-validation) a JSONSchema validator for Postgres which might help

**Comment by lxblvs** [comment]
So the solution now is to change the field to jsonb? Mkay.

**Comment by Shalashtein** [comment]
This issue is making materialized views in hasura a pain

**Comment by danieljharvey** [comment]
Hello! As of version `2.29` we have built-in support for Postgres arrays.

You can now use native arrays to insert array items, like this:

```graphql
mutation insert_author {
  insert_author (
    objects: [
      {
        name: "Ash",
        emails: ["ash@ash.com", "ash123@ash.com"]
      }
    ]
  ) {
    affected_rows
    returning {
      id
      name
      emails
    }
  }
}
```

You can also use `_contains` and `_contained_in` to filter array fields.

More information can be found [in the docs](https://hasura.io/docs/latest/mutations/postgres/insert/#insert-an-object-with-an-array-field).

**Comment by marionschleifer** [comment]
Hey folks, we will give a quick update on supporting Postgres arrays in the community call this Thursday, 27th July at 9am PT, along with many other community updates 🙌  if you're interested, you can sign up [here](https://hasura.io/community/community-call/).

**Comment by bensoutendijk** [comment]
> Hey folks, we will give a quick update on supporting Postgres arrays in the community call this Thursday, 27th July at 9am PT, along with many other community updates 🙌 if you're interested, you can sign up [here](https://hasura.io/community/community-call/).

Can someone give a summary of the update from this call?
